lnwallet/btcwallet/btcwallet.go

```diff
@@ -1199,33 +1201,102 @@ func (b *BtcWallet) ListUnspentWitness(minConfs, maxConfs int32,
 // already published to the network (either in the mempool or chain) no error
 // will be returned.
 func (b *BtcWallet) PublishTransaction(tx *wire.MsgTx, label string) error {
-	if err := b.wallet.PublishTransaction(tx, label); err != nil {
+	// handleErr is a helper closure that handles the error from
+	// PublishTransaction and TestMempoolAccept.
+	handleErr := func(err error) error {
 		// If we failed to publish the transaction, check whether we
 		// got an error of known type.
-		switch err.(type) {
+		switch {
 		// If the wallet reports a double spend, convert it to our
 		// internal ErrDoubleSpend and return.
-		case *base.ErrDoubleSpend:
+		case lnutils.ErrorAs[*base.ErrDoubleSpend](err):
 			return lnwallet.ErrDoubleSpend
 
 		// If the wallet reports a replacement error, return
 		// ErrDoubleSpend, as we currently are never attempting to
 		// replace transactions.
-		case *base.ErrReplacement:
+		case lnutils.ErrorAs[*base.ErrReplacement](err):
 			return lnwallet.ErrDoubleSpend
 
-		// If the wallet reports that fee requirements for accepting the
-		// tx into mempool are not met, convert it to our internal
+		// If the wallet reports that fee requirements for accepting
+		// the tx into mempool are not met, convert it to our internal
 		// ErrMempoolFee and return.
-		case *base.ErrMempoolFee:
+		case lnutils.ErrorAs[*base.ErrMempoolFee](err):
 			return fmt.Errorf("%w: %v", lnwallet.ErrMempoolFee,
 				err.Error())
-
-		default:
-			return err
 		}
+
+		return err
 	}
-	return nil
+
+	// For neutrino backend there's no mempool, so we return early by
+	// publishing the transaction.
+	if b.chain.BackEnd() == "neutrino" {
+		err := b.wallet.PublishTransaction(tx, label)
+
+		return handleErr(err)
+	}
+
+	// For non-neutrino nodes, we will first check whether the transaction
+	// can be accepted by the mempool.
+	// Use a max feerate of 0 means the default value will be used when
+	// testing mempool acceptance. The default max feerate is 0.10 BTC/kvb,
+	// or 10,000 sat/vb.
+	results, err := b.chain.TestMempoolAccept([]*wire.MsgTx{tx}, 0)
+	if err != nil {
+		return err
```

Comment:
I think we can improve the test slightly by having this return a new wrapper error. Then we can assert against this particular error to ensure that it was rejected by `TestMempoolAccept` rather than an error later once we actually go to publish.

Reply:
Hmm, I guess `handleError` indirectly achieves this.

Reply:
yeah exactly